src/libPMacc/include/particles/operations/CountParticles.hpp

```diff
@@ -31,72 +30,163 @@
 #include "particles/particleFilter/PositionFilter.hpp"
 #include "nvidia/atomic.hpp"
 #include "memory/shared/Allocate.hpp"
-
+#include "traits/GetNumWorkers.hpp"
+#include "mappings/threads/ForEachIdx.hpp"
+#include "mappings/threads/IdxConfig.hpp"
 
 
 namespace PMacc
 {
 
-/* count particles in an area
- * is not optimized, it checks any partcile position if its realy a particle
+/* count particles
+ *
+ * it is allowed to call this kernel on frames with holes (without calling fillAllGAps before)
+ *
+ * @tparam T_numWorkers number of workers
  */
+template< uint32_t T_numWorkers >
 struct KernelCountParticles
 {
-    template<class PBox, class Filter, class Mapping>
-    DINLINE void operator()(
-        PBox pb,
+    /** count particles
+     *
+     * @tparam T_PBox PMacc::ParticlesBox, particle box type
+     * @tparam T_Filter functor to filter particles
+     * @tparam T_Mapping supercell mapper functor type
+     *
+     * @param pb particle memory
+     * @param gCounter pointer for the result
+     * @param filter functor to filter particles those should be counted
+     * @param mapper functor to map a block to a supercell
+     */
+    template<
+        typename T_PBox,
+        typename T_Filter,
+        typename T_Mapping
+    >
+    DINLINE void operator( )(
+        T_PBox & pb,
         uint64_cu* gCounter,
```

Comment:
not sure why we don't design this kernel as a pretty `uint64_t &`, that's messy downstream on all places